### PR TITLE
Reactive extension WKWebView navigation delegate

### PR DESCRIPTION
**Pull request by gringoireDM**
This PR add a new delegate proxy for WKNavigationDelegate for WKWebView.
Losing UIWebView delegate proxy, this PR now offers a replacement.

Please note that WKWebView has multiple delegates. This PR offers a proxy just for the WKNavigationDelegate. Other delegates will be subject of future PRs

### CODE REVIEW

File: RxCocoa/iOS/WKWebView+Rx.swift [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    public var navigationDelegate: DelegateProxy<WKWebView, WKNavigationDelegate> {
+        RxWKNavigationDelegateProxy.proxy(for: base)
+    }
+    
+    /// Reactive wrapper for `navigationDelegate` message.
+    public var didCommit: Observable<WKNavigation> {
+        navigationDelegate
+            .methodInvoked(#selector(WKNavigationDelegate.webView(_:didCommit:)))
+            .map { a in try castOrThrow(WKNavigation.self, a[1]) }
+
```
Comment (freak4pc) [inline RxCocoa/iOS/WKWebView+Rx.swift]:
```suggestion
```

File: RxCocoa/iOS/WKWebView+Rx.swift [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    /// Reactive wrapper for `navigationDelegate` message.
+    public var didFailLoad: Observable<(WKNavigation, Error)> {
+        navigationDelegate
+            .methodInvoked(#selector(WKNavigationDelegate.webView(_:didFail:withError:)))
+            .map { a in
+                (
+                    try castOrThrow(WKNavigation.self, a[1]),
+                    try castOrThrow(Error.self, a[2])
+                )
+        }
```
Comment (freak4pc) [inline RxCocoa/iOS/WKWebView+Rx.swift]: Keep curly braces aligned to opening please

```suggestion
            }
```

File: Tests/RxCocoaTests/DelegateProxyTest+WebKit.swift [hunk 3]
```diff
@@ -0,0 +1,49 @@
+//
+//  DelegateProxyTest+WebKit.swift
+//  Tests
+//
+//  Created by Giuseppe Lanza on 14/02/2020.
+//  Copyright © 2020 Krunoslav Zaher. All rights reserved.
+//
+
+#if os(iOS) || os(macOS)
```
Comment (freak4pc) [inline Tests/RxCocoaTests/DelegateProxyTest+WebKit.swift]: I think you don't need this entire file, the delegate proxy is already tested on UIKit level. (For example there isn't a specific `DelegateProxyTest+UITableView`). 

As long as you have a `WKWebViewTests`, that should suffice:]

File: Tests/RxCocoaTests/DelegateProxyTest.swift [hunk 4]
```diff
@@ -446,6 +447,9 @@ extension DelegateProxyTest {
         autoreleasepool {
             control = nil
         }
+        
+        wait(for: [expectation], timeout: 1)
```
Comment (freak4pc) [inline Tests/RxCocoaTests/DelegateProxyTest.swift]: why is this needed? seems like the tests pass without it as well

File: Tests/RxCocoaTests/DelegateProxyTest.swift [hunk 5]
```diff
@@ -770,3 +774,13 @@ extension MockTestDelegateProtocol
...
+
+#if os(iOS) || os(macOS)
+import WebKit
+
+@available(iOS 10.0, OSXApplicationExtension 10.10, *)
+extension MockTestDelegateProtocol
+    : WKNavigationDelegate
+{
+}
+#endif
```
Comment (freak4pc) [inline Tests/RxCocoaTests/DelegateProxyTest.swift]: Same, probably not needed

File: Tests/RxCocoaTests/WKWebView+RxTests.swift [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+import XCTest
+
+// Any WKNavigation object manually created on dealloc crashes the program.
+// This class overrides the deinit method of the WKNavition to avoid crashes.
+@available(iOS 10.0, OSXApplicationExtension 10.10, *)
+private class SafeWKNavigation: WKNavigation {
+    static func toggleSafeDealloc() {
+        guard let current_original = class_getInstanceMethod(SafeWKNavigation.self, NSSelectorFromString("dealloc")),
+            let current_swizzled = class_getInstanceMethod(SafeWKNavigation.self, #selector(nonCrashingDealloc)) else { return }
+        method_exchangeImplementations(current_original, current_swizzled)
```
Comment (freak4pc) [inline Tests/RxCocoaTests/WKWebView+RxTests.swift]:
```suggestion
        guard let current_original = class_getInstanceMethod(SafeWKNavigation.self, NSSelectorFromString("dealloc")),
              let current_swizzled = class_getInstanceMethod(SafeWKNavigation.self, #selector(nonCrashingDealloc)) else { return }
        method_exchangeImplementations(current_original, current_swizzled)
```

Comment (freak4pc) [reply Tests/RxCocoaTests/WKWebView+RxTests.swift]: Could you also move the helpers below the test class itself under a mark? e.g.

```swift
final class WKWebViewTests { ... }

// MARK: - Test Helpers

private class SafeWKNavigation: WKNavigation  { }
```

File: Tests/RxCocoaTests/WKWebView+RxTests.swift [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    }
+    
+    @objc func nonCrashingDealloc() { }
+}
+
+extension String: Error {}
```
Comment (freak4pc) [inline Tests/RxCocoaTests/WKWebView+RxTests.swift]: I prefer not using this trick. Can you wrap the error in something private for this test using an actual concrete type? even an enum case with a String associated value would be fine, or something similar. This would make all Strings in all tests conform to Error.

**Comment by freak4pc** [comment]
P.S. @gringoireDM this also needs a CHANGELOG entry, please :)

File: Tests/RxCocoaTests/DelegateProxyTest+WebKit.swift (hunk 3 above)
Comment (gringoireDM) [reply Tests/RxCocoaTests/DelegateProxyTest+WebKit.swift]: WKWebView behaves differently in dealloc. Dealloc doesn't happen immediately as expected so I preferred to test as it was tested for UIWebView before.

File: Tests/RxCocoaTests/DelegateProxyTest.swift (hunk 4 above)
Comment (gringoireDM) [reply Tests/RxCocoaTests/DelegateProxyTest.swift]: WKWebView is not deallocated immediately. Internally there is something else that retains the WKWebView for a little longer than expected:

https://stackoverflow.com/questions/39349911/memory-leak-in-wkwebview-instantiation

File: Tests/RxCocoaTests/DelegateProxyTest.swift (hunk 5 above)
Comment (gringoireDM) [reply Tests/RxCocoaTests/DelegateProxyTest.swift]: same as https://github.com/ReactiveX/RxSwift/pull/2144#discussion_r391470310

**Comment by gringoireDM** [comment]
@freak4pc is there a swiftFormat step I skipped? If not, should we add one?

File: Tests/RxCocoaTests/DelegateProxyTest+WebKit.swift (hunk 3 above)
Comment (freak4pc) [reply Tests/RxCocoaTests/DelegateProxyTest+WebKit.swift]: Right, but you are testing a lot of non-dealloc behavior (e.g. copying the "doThisTest" stuff etc). Maybe worth just testing the specific deallocation? Or is there any reason to test the surrounding?

**Comment by freak4pc** [comment]
@gringoireDM We don't have SwiftFormat ATM. Anything specific you're referring to adding this for? Happy to accept a PR if it's anything helpful. 

Thanks !

File: Tests/RxCocoaTests/DelegateProxyTest.swift (hunk 5 above)
Comment (freak4pc) [reply Tests/RxCocoaTests/DelegateProxyTest.swift]: Still somewhat confused - if you have tests in a separate file, maybe this conformance should be there, and not in the general Delegate Proxy Test file.

File: Tests/RxCocoaTests/DelegateProxyTest+WebKit.swift (hunk 3 above)
Comment (gringoireDM) [reply Tests/RxCocoaTests/DelegateProxyTest+WebKit.swift]: Code coverage mainly. It's a new file, with a new Delegate proxy. Without this test `RxWKNavigationDelegateProxy` won't be covered and while we can agree that it follows the same pattern as others we can't be sure about future refactoring and human errors.

**Comment by gringoireDM** [comment]
@freak4pc a lot of changes for requests comes for spacings and listing stuff we can uniform using swiftFormat rules so we won't need to comment and fix those anymore in PRs. 

Anyway, it's something for a separate discussion not related to this PR specifically at the moment.

**Comment by freak4pc** [comment]
Happy to. Don’t have a lot of experience with customizing it and Xcode’s default spacing is quite lacking.

**Comment by gringoireDM** [comment]
I'll prepare one asap

File: Tests/RxCocoaTests/DelegateProxyTest+WebKit.swift (hunk 3 above)
Comment (freak4pc) [reply Tests/RxCocoaTests/DelegateProxyTest+WebKit.swift]: Alright. I don't mind much about coverage in this case (not everything has to be covered), but I don't mind the extra test.

File: Tests/RxCocoaTests/DelegateProxyTest.swift (hunk 5 above)
Comment (freak4pc) [reply Tests/RxCocoaTests/DelegateProxyTest.swift]: So can we move this to **DelegateProxyTest+WebKit.swift** ?

Comment (gringoireDM) [reply Tests/RxCocoaTests/DelegateProxyTest.swift]: sure makes sense.

I felt safer developing the delegate proxy for the first time having a test confirming I did it right, you know what I mean?

Comment (freak4pc) [reply Tests/RxCocoaTests/DelegateProxyTest.swift]: Yup, that's fair. :)
